rfcs/032-lineage-backend.md

```diff
@@ -0,0 +1,97 @@
+- Feature Name: lineage-backend
+- Start Date: 2021-03-31
+- RFC PR: [amundsen-io/rfcs#32](https://github.com/amundsen-io/rfcs/pull/32)
+- Amundsen Issue: [amundsen-io/amundsen#0000](https://github.com/amundsen-io/amundsen/issues/0000) (leave this empty for now)
+
+# Amundsen Lineage - Backend
+
+## Summary
+
+Currently, we link to the 3rd Party lineage tool in order to show table and column lineage for tables in Amundsen. 
+The goal of this project is to have a lineage graph that is native to Amundsen so that users can find table and column lineage in one platform.
+
+## Motivation
+
+At the moment we are relying completely on any 3rd party lineage tool to show the table and column lineage in Amundsen. 
+Not everyone has access to such a 3rd party tool and want to have lineage information available natively in Amundsen.
+
+ref: https://github.com/amundsen-io/rfcs/blob/master/rfcs/025-lineage-stage-0.md
+
+## Guide-level Explanation (aka Product Details)
+
+No new concepts/definitions are being introduced as a part of this RFC. 
+
+Databuilder already has the table lineage model, which creates an upstream/downstream relation to adding to the Neo4j graph. 
+Column lineage model however still needs to be developed as a part of this RFC.
```

Comment:
It would be reasonable to define this in the RFC. That said, given there's already table lineage, and the metadata response formats are defined, I think the solution space is small enough it's probably OK to figure it out in the implementation

Reply:
I agree, because we have started to define neo4j queries based up on pervious RFC and also lineage model available into amundsen-common. though these are not optimized queries and not completed but I was planning to put queries on this line. 

```
    def get_lineage(self, *,
                    id: str,
                    resource_type: ResourceType, direction: str, depth: int) -> Lineage:

        get_both_lineage_query = textwrap.dedent(u"""
        MATCH (down_parent:Table)<-[downstream_len:DOWNSTREAM*..{depth_key}]-(child:Table {{key: $query_key }})-[upstream_len:UPSTREAM*..{depth_key}]->(up_parent:Table)
        WITH
        child.key as child_key
        ,collect(distinct{{level:LENGTH(upstream_len),source:'hive',key:up_parent.key}}) AS upstream_entities
        ,collect(distinct{{level:LENGTH(downstream_len),source:'hive',key:down_parent.key}}) AS downstream_entities
        RETURN
        collect({{
        key:child_key,direction:"both",depth:1
        ,upstream_entities:upstream_entities
        ,downstream_entities:downstream_entities
        }}) AS lineageOutput
        """).format(depth_key=depth)

        get_upstream_lineage_query = textwrap.dedent(u"""
        MATCH (child:Table {{key: $query_key }})-[upstream_len:UPSTREAM*..{depth_key}]->(up_parent:Table)
        WITH
        child.key as child_key
        ,collect(distinct{{level:LENGTH(upstream_len),source:'hive',key:up_parent.key}}) AS upstream_entities
        RETURN
        collect({{
        key:child_key,direction:"upstream",depth:1
        ,upstream_entities:upstream_entities
        }}) AS lineageOutput
        """).format(depth_key=depth)

        get_downstream_lineage_query = textwrap.dedent(u"""
        MATCH (down_parent:Table)<-[downstream_len:DOWNSTREAM*..{depth_key}]-(child:Table {{key: $query_key }})
        WITH
        child.key as child_key
        ,collect(distinct{{level:LENGTH(downstream_len),source:'hive',key:down_parent.key}}) AS downstream_entities
        RETURN
        collect({{
        key:child_key,direction:"downstream",depth:1
        ,downstream_entities:downstream_entities
        }}) AS lineageOutput
        """).format(depth_key=depth)

        if direction == 'upstream':
            records = self._execute_cypher_query(statement=get_upstream_lineage_query,
                                                 param_dict={'query_key': id})
        elif direction == 'downstream':
            records = self._execute_cypher_query(statement=get_downstream_lineage_query,
                                                 param_dict={'query_key': id})
        else:
            records = self._execute_cypher_query(statement=get_both_lineage_query,
                                                 param_dict={'query_key': id})

        result = records.single()['lineageOutput'][0]
        return result
```